File: viewer/frontend/src/lib/api/monitor-data.ts
import { z } from "zod";

import { requestJson } from "@/lib/api/client";

export const monitorDataSchema = z.object({
  jobId: z.string(),
  nodePath: z.string(),
  preset: z.string().nullable().optional(),
  dataset: z.string().nullable(),
  logDir: z.string().nullable(),
  scalarSeries: z.array(
    z.object({
      tag: z.string(),
      label: z.string(),
      points: z.array(
        z.object({
          step: z.number(),
          wallTime: z.number(),
          value: z.number(),
        }),
      ),
    }),
  ),
  histograms: z.array(
    z.object({
      tag: z.string(),
      step: z.number(),
      wallTime: z.number(),
      buckets: z.array(
        z.object({
          left: z.number(),
          right: z.number(),
          count: z.number(),
        }),
      ),
    }),
  ),
  images: z.array(
    z.object({
      tag: z.string(),
      step: z.number(),
      wallTime: z.number(),
      mimeType: z.string(),
      dataUrl: z.string(),
    }),
  ),
});

export const parameterChannelStatusSchema = z.object({
  status: z.enum(["updated", "unchanged", "missing", "unknown"]),
  metric: z.string().nullable().optional(),
  lastStep: z.number().nullable().optional(),
  observedPoints: z.number(),
});

export const parameterNodeStatusSchema = z.object({
  nodePath: z.string(),
  weights: parameterChannelStatusSchema,
  bias: parameterChannelStatusSchema,
});

export const parameterStatusSchema = z.object({
  sourceId: z.string(),
  preset: z.string().nullable().optional(),
  dataset: z.string().nullable().optional(),
  logDir: z.string().nullable().optional(),
  nodes: z.array(parameterNodeStatusSchema),
});

export const logParameterStatusSchema = z.object({
  runs: z.array(parameterStatusSchema),
});

export type MonitorData = z.infer<typeof monitorDataSchema>;
export type ParameterActivityStatus = z.infer<
  typeof parameterChannelStatusSchema
>["status"];
export type ParameterChannelStatus = z.infer<typeof parameterChannelStatusSchema>;
export type ParameterNodeStatus = z.infer<typeof parameterNodeStatusSchema>;
export type ParameterStatus = z.infer<typeof parameterStatusSchema>;
export type LogParameterStatusResponse = z.infer<typeof logParameterStatusSchema>;

export function fetchMonitorData(input: {
  jobId: string;
  nodePath: string;
  preset?: string;
  dataset?: string;
}) {
  const params = new URLSearchParams({ nodePath: input.nodePath });
  if (input.preset) {
    params.set("preset", input.preset);
  }
  if (input.dataset) {
    params.set("dataset", input.dataset);
  }
  return requestJson(
    `/training/jobs/${input.jobId}/monitor-data?${params.toString()}`,
    monitorDataSchema,
  );
}

export function fetchMonitorParameterStatus(input: {
  jobId: string;
  preset?: string;
  dataset?: string;
}) {
  const params = new URLSearchParams();
  if (input.preset) {
    params.set("preset", input.preset);
  }
  if (input.dataset) {
    params.set("dataset", input.dataset);
  }
  const query = params.toString();
  return requestJson(
    `/training/jobs/${input.jobId}/monitor-parameter-status${query ? `?${query}` : ""}`,
    parameterStatusSchema,
  );
}

export function fetchLogRunMonitorData(input: {
  runId: string;
  nodePath: string;
}) {
  const params = new URLSearchParams({ nodePath: input.nodePath });
  return requestJson(
    `/logs/runs/${encodeURIComponent(input.runId)}/monitor-data?${params.toString()}`,
    monitorDataSchema,
  );
}

export function fetchLogParameterStatus(input: { runIds: string[] }) {
  return requestJson("/logs/parameter-status", logParameterStatusSchema, {
    method: "POST",
    body: JSON.stringify(input),
  });
}

File: workbench/backend/schemas/_monitor_data.py
from __future__ import annotations

from typing import Literal

from pydantic import Field

from workbench.backend.schemas._base import ApiResponseModel, BoundedIdentifier
from workbench.backend.schemas._limits import MAX_LOG_REQUEST_RUN_IDS


def _is_none(value: object) -> bool:
    return value is None


class ScalarPointResponse(ApiResponseModel):
    step: int
    wallTime: float
    value: float


class ScalarSeriesResponse(ApiResponseModel):
    tag: str
    label: str
    points: list[ScalarPointResponse]
    sourceItemCount: int | None = Field(default=None, exclude_if=_is_none)
    returnedItemCount: int | None = Field(default=None, exclude_if=_is_none)
    truncated: bool | None = Field(default=None, exclude_if=_is_none)
    truncationReason: str | None = Field(default=None, exclude_if=_is_none)


class HistogramBucketResponse(ApiResponseModel):
    left: float
    right: float
    count: float


class HistogramResponse(ApiResponseModel):
    tag: str
    step: int
    wallTime: float
    buckets: list[HistogramBucketResponse]
    sourceItemCount: int | None = Field(default=None, exclude_if=_is_none)
    returnedItemCount: int | None = Field(default=None, exclude_if=_is_none)
    truncated: bool | None = Field(default=None, exclude_if=_is_none)
    truncationReason: str | None = Field(default=None, exclude_if=_is_none)


class ImageResponse(ApiResponseModel):
    tag: str
    step: int
    wallTime: float
    mimeType: str
    dataUrl: str
    eventBytes: int | None = Field(default=None, exclude_if=_is_none)
    sourceItemCount: int | None = Field(default=None, exclude_if=_is_none)
    returnedItemCount: int | None = Field(default=None, exclude_if=_is_none)
    truncated: bool | None = Field(default=None, exclude_if=_is_none)
    truncationReason: str | None = Field(default=None, exclude_if=_is_none)


class MonitorDataResponse(ApiResponseModel):
    jobId: str
    nodePath: str
    preset: str | None = None
    dataset: str | None = None
    logDir: str | None = None
    eventBytes: int | None = Field(default=None, exclude_if=_is_none)
    skippedEventFiles: int | None = Field(default=None, exclude_if=_is_none)
    truncated: bool | None = Field(default=None, exclude_if=_is_none)
    truncationReason: str | None = Field(default=None, exclude_if=_is_none)
    sourceItemCount: int | None = Field(default=None, exclude_if=_is_none)
    returnedItemCount: int | None = Field(default=None, exclude_if=_is_none)
    scalarSeries: list[ScalarSeriesResponse]
    histograms: list[HistogramResponse]
    images: list[ImageResponse]


ParameterActivityStatus = Literal["updated", "unchanged", "missing", "unknown"]


class ParameterChannelStatusResponse(ApiResponseModel):
    status: ParameterActivityStatus
    metric: str | None = None
    lastStep: int | None = None
    observedPoints: int


class ParameterNodeStatusResponse(ApiResponseModel):
    nodePath: str
    weights: ParameterChannelStatusResponse
    bias: ParameterChannelStatusResponse


class ParameterStatusResponse(ApiResponseModel):
    sourceId: str
    preset: str | None = None
    dataset: str | None = None
    logDir: str | None = None
    eventBytes: int | None = Field(default=None, exclude_if=_is_none)
    skippedEventFiles: int | None = Field(default=None, exclude_if=_is_none)
    truncated: bool | None = Field(default=None, exclude_if=_is_none)
    truncationReason: str | None = Field(default=None, exclude_if=_is_none)
    sourceItemCount: int | None = Field(default=None, exclude_if=_is_none)
    returnedItemCount: int | None = Field(default=None, exclude_if=_is_none)
    nodes: list[ParameterNodeStatusResponse]


class LogParameterStatusRequest(ApiResponseModel):
    runIds: list[BoundedIdentifier] = Field(
        default_factory=list,
        max_length=MAX_LOG_REQUEST_RUN_IDS,
    )


class LogParameterStatusResponse(ApiResponseModel):
    runs: list[ParameterStatusResponse]
